### PR TITLE
42733: Restore CAS auth functionality

### PR DESCRIPTION
**Pull request by Rillke**
- ~Allow `phpCAS` to `unset` variables~
- do not use deprecated `phpCAS::setDebug()`
- fix call to `phpCAS::client()` by providing ILIAS base URL
- Add more debug logging

https://mantis.ilias.de/view.php?id=42733

### CODE REVIEW

**Comment by Rillke** [comment]
> copyright is not as expected in Services/CAS/classes/class.ilAuthProviderCAS.php
> copyright is not as expected in src/HTTP/Wrapper/SuperGlobalDropInReplacement.php

yeah, just tell me what's expected and how to fix it. i suppose i will not change any copyright header.

File: src/HTTP/Wrapper/SuperGlobalDropInReplacement.php [hunk 1]
```diff
@@ -54,6 +54,13 @@ public function offsetSet($offset, $value): void
...
     {
-        throw new LogicException("Modifying global Request-Array such as \$_GET is not allowed!");
+        if ($offset === 'ticket') {
+            // Allow phpCAS to unset the 'ticket'
+            // they do this in CAS/Client.php#L1070
+            // unset($_GET['ticket']);
+            parent::offsetUnset($offset);
+        } else {
+            throw new LogicException("Modifying global Request-Array such as \$_GET is not allowed!");
+        }
```
Comment (mjansenDatabay) [inline src/HTTP/Wrapper/SuperGlobalDropInReplacement.php]: Please revert this change and follow the instructions here: https://github.com/ILIAS-eLearning/ILIAS/tree/release_8/src/HTTP#dropinreplacements

File: Services/CAS/classes/class.ilAuthProviderCAS.php [hunk 2]
```diff
@@ -38,29 +67,54 @@ protected function getSettings(): ilCASSettings
...
+            // Though, you may need to for debugging other issues.
+            phpCAS::setVerbose(false);
+            $this->getLogger()->debug('Create client... ');
             phpCAS::client(
                 CAS_VERSION_2_0,
                 $this->getSettings()->getServer(),
                 $this->getSettings()->getPort(),
-                $this->getSettings()->getUri()
+                $this->getSettings()->getUri(),
+                $baseUrl
```
Comment (mjansenDatabay) [inline Services/CAS/classes/class.ilAuthProviderCAS.php]: Can't we use `ilUtil::_getHttpPath()` instead?

Comment (Rillke) [reply Services/CAS/classes/class.ilAuthProviderCAS.php]: I'm not sure if it is an officially supported feature of ILIAS to work under multiple domain names and if forcing people to a configured domain name (which does not necessarily have to be the one the user is currently on) causes issues anywhere. But it's probably the way with the smallest surface for bad actors.

Comment (mjansenDatabay) [reply Services/CAS/classes/class.ilAuthProviderCAS.php]: > if it is an officially supported feature of ILIAS to work under multiple domain names

It's not.

I suggest using the above method to avoid introducing another “truth” regarding the “HTTP path”.

File: Services/CAS/classes/class.ilAuthProviderCAS.php [hunk 3]
```diff
@@ -38,29 +67,54 @@ protected function getSettings(): ilCASSettings
...
+            // entire CAS-call-stack.
+            //
+            // $this->logPath = tempnam(sys_get_temp_dir(), 'phpCAS');
+            // $this->logger = new Logger('name');
+            // $handler = new StreamHandler($this->logPath);
+            // $format = "%message%\n";
+            // $formatter = new LineFormatter($format);
+            // $handler->setFormatter($formatter);
+            // $this->logger->pushHandler($handler);
+            // phpCAS::setLogger($this->logger);
```
Comment (mjansenDatabay) [inline Services/CAS/classes/class.ilAuthProviderCAS.php]: I suggest injecting the PSR-compliant logger wrapped by the ILIAS logger, e.g.:

```php
$ilias_logger = $DIC->logger()->auth();

phpCAS::setLogger($ilias_logger->getLogger());
```

File: Services/CAS/classes/class.ilAuthProviderCAS.php [hunk 4]
```diff
@@ -22,14 +26,37 @@
  */
 class ilAuthProviderCAS extends ilAuthProvider
 {
-    private ilCASSettings $settings;
+    // private ilCASSettings $settings;
```
Comment (Rillke) [inline Services/CAS/classes/class.ilAuthProviderCAS.php]: this was unintended.

File: Services/CAS/classes/class.ilAuthProviderCAS.php (hunk 3 above)
Comment (mjansenDatabay) [reply Services/CAS/classes/class.ilAuthProviderCAS.php]: or even better:

```php
phpCAS::setLogger($this->getLogger()->getLogger());
```

**Review by mjansenDatabay — CHANGES_REQUESTED** [review]
As suggested in the inline comments: Please pass the PSR-compliant logger to the CAS-library.

File: Services/CAS/classes/class.ilAuthProviderCAS.php (hunk 3 above)
Comment (Rillke) [reply Services/CAS/classes/class.ilAuthProviderCAS.php]: @mjansenDatabay I suppose it should be only one `->getLogger()`.

What I noticed is that CAS will still log a lot with log level INFO; user details that might be unwanted in log files, so I will check if phpCAS would somehow allow to be less verbose by default.

```
[0dc17] [2024-11-19 19:27:15.474161] client_auth.INFO: phpCAS::trace:468 47AC .START (2024-11-19 19:27:15) phpCAS-1.6.1 ****************** [CAS.php:468]  
[0dc17] [2024-11-19 19:27:15.474320] client_auth.INFO: phpCAS::traceBegin:352 47AC .=> phpCAS::client('2.0', 'shib.rz.example.com', 443, 'idp/profile/cas', 'https://dev.ilias.example.com') [class.ilAuthProviderCAS.php:56]  
[0dc17] [2024-11-19 19:27:15.474460] client_auth.INFO: phpCAS::traceBegin:963 47AC .|    => CAS_Client::__construct('2.0', false, 'shib.rz.example.com', 443, 'idp/profile/cas', 'https://dev.ilias.example.com', true, CAS_Session_PhpSession) [CAS.php:369]  
[0dc17] [2024-11-19 19:27:15.474494] client_auth.INFO: phpCAS::trace:982 47AC .|    |    Session is not authenticated [Client.php:982]  
[0dc17] [2024-11-19 19:27:15.474528] client_auth.INFO: phpCAS::traceEnd:1080 47AC .|    <= ''  
[0dc17] [2024-11-19 19:27:15.474545] client_auth.INFO: phpCAS::traceEnd:375 47AC .<= ''  
[0dc17] [2024-11-19 19:27:15.474562] client_auth.INFO: phpCAS::traceBegin:1702 47AC .=> phpCAS::setNoCasServerValidation() [class.ilAuthProviderCAS.php:59]  
[0dc17] [2024-11-19 19:27:15.474581] client_auth.INFO: phpCAS::trace:1705 47AC .|    You have configured no validation of the legitimacy of the cas server. This is not recommended for production use. [CAS.php:1705]  
[0dc17] [2024-11-19 19:27:15.474596] client_auth.INFO: phpCAS::traceEnd:1707 47AC .<= ''  
[0dc17] [2024-11-19 19:27:15.474613] client_auth.INFO: phpCAS::traceBegin:1159 47AC .=> phpCAS::forceAuthentication() [class.ilAuthProviderCAS.php:61]  
[0dc17] [2024-11-19 19:27:15.474630] client_auth.INFO: phpCAS::traceBegin:1486 47AC .|    => CAS_Client::forceAuthentication() [CAS.php:1161]  
[0dc17] [2024-11-19 19:27:15.474646] client_auth.INFO: phpCAS::traceBegin:1592 47AC .|    |    => CAS_Client::isAuthenticated() [Client.php:1488]  
...
```

and hundreds of lines more. This is great for log level debug but not for log level info. So please do not yet merge.

**Comment by Rillke** [comment]
> As suggested in the inline comments: Please pass the PSR-compliant logger to the CAS-library.

done

Please don't merge yet. Still looking for a way to make phpCAS not sending trace to log level INFO.

**Comment by Rillke** [comment]
@mjansenDatabay it seems phpCAS has hardcoded info-level logging for traces:

See https://github.com/apereo/phpCAS/blob/690d431005e4f88651499458469536201623b057/source/CAS.php#L568

The output looks like:

```
[0dc17] [2024-11-19 19:27:15.474161] client_auth.INFO: phpCAS::trace:468 47AC .START (2024-11-19 19:27:15) phpCAS-1.6.1 ****************** [CAS.php:468]  
[0dc17] [2024-11-19 19:27:15.474320] client_auth.INFO: phpCAS::traceBegin:352 47AC .=> phpCAS::client('2.0', 'shib.rz.example.com', 443, 'idp/profile/cas', 'https://dev.ilias.example.com') [class.ilAuthProviderCAS.php:56]  
[0dc17] [2024-11-19 19:27:15.474460] client_auth.INFO: phpCAS::traceBegin:963 47AC .|    => CAS_Client::__construct('2.0', false, 'shib.rz.example.com', 443, 'idp/profile/cas', 'https://dev.ilias.example.com', true, CAS_Session_PhpSession) [CAS.php:369]  
[0dc17] [2024-11-19 19:27:15.474494] client_auth.INFO: phpCAS::trace:982 47AC .|    |    Session is not authenticated [Client.php:982]  
[0dc17] [2024-11-19 19:27:15.474528] client_auth.INFO: phpCAS::traceEnd:1080 47AC .|    <= ''  
[0dc17] [2024-11-19 19:27:15.474545] client_auth.INFO: phpCAS::traceEnd:375 47AC .<= ''  
[0dc17] [2024-11-19 19:27:15.474562] client_auth.INFO: phpCAS::traceBegin:1702 47AC .=> phpCAS::setNoCasServerValidation() [class.ilAuthProviderCAS.php:59]  
[0dc17] [2024-11-19 19:27:15.474581] client_auth.INFO: phpCAS::trace:1705 47AC .|    You have configured no validation of the legitimacy of the cas server. This is not recommended for production use. [CAS.php:1705]  
[0dc17] [2024-11-19 19:27:15.474596] client_auth.INFO: phpCAS::traceEnd:1707 47AC .<= ''  
[0dc17] [2024-11-19 19:27:15.474613] client_auth.INFO: phpCAS::traceBegin:1159 47AC .=> phpCAS::forceAuthentication() [class.ilAuthProviderCAS.php:61]  
[0dc17] [2024-11-19 19:27:15.474630] client_auth.INFO: phpCAS::traceBegin:1486 47AC .|    => CAS_Client::forceAuthentication() [CAS.php:1161]  
[0dc17] [2024-11-19 19:27:15.474646] client_auth.INFO: phpCAS::traceBegin:1592 47AC .|    |    => CAS_Client::isAuthenticated() [Client.php:1488]  
...
```

We'd better leave `phpCAS::setLogger($this->getLogger());` as a comment?

**Comment by mjansenDatabay** [comment]
Hi @Rillke ,

thanks a lot for your changes. I agree, the logging is to verbose. We should better leave this as a comment.

After this (Update: You already changed it), I will squash the commits, merge the PR, and cherry-pick the commit to all maintained branches (incl. ILIAS 10).

In general:

What worries me a little is the state of `phpCAS` described here: https://github.com/apereo/phpCAS/tree/690d431005e4f88651499458469536201623b057?tab=readme-ov-file#project-is-unmaintained .
While the library was still accepted for ILIAS 10 (see: https://github.com/ILIAS-eLearning/ILIAS/pull/6765), I doubt that we will be able to do so again for ILIAS 11.
So what we need is an alternative library, preferably within the next three weeks. In December all libraries for ILIAS 11 have to be proposed (again) as a pull request, see: https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/vendor/README.md#re-evaluating-dependencies-on-major-updates-of-ilias . What we also need is someone who uses/has knowledge in CAS and can push the topic for 11 (together with us). I don't have any “shares” in CAS myself.

If you (or someone else who reads this and needs CAS for authentication in ILIAS) know(s) of alternatives here and want(s) to adapt the code in `ilAuthProviderCAS` even for ILIAS 11 to a newly proposed library via pull request, this would, of course, be enormously helpful for obtaining CAS authentication.

Best regards,
@mjansenDatabay

**Review by mjansenDatabay — APPROVED** [review]

**Comment by Rillke** [comment]
Thanks for the review and the merge. You know, we would pay for such a service, too.

> So what we need is an alternative library, preferably within the next three weeks. In December all libraries for ILIAS 11 have to be proposed (again) as a pull request, see: https://github.com/ILIAS-eLearning/ILIAS/blob/trunk/vendor/README.md#re-evaluating-dependencies-on-major-updates-of-ilias . What we also need is someone who uses/has knowledge in CAS and can push the topic for 11 (together with us). I don't have any “shares” in CAS myself.

We are currently switching from CAS Server with CAS protocol to Shibboleth Server with CAS protocol (because our IDM team does not want to maintain two SSO software stacks and Shibboleth is more or less mandated by DFN). We will probalbly switch to Shibboleth Server with OIDC soon.

> If you (or someone else who reads this and needs CAS for authentication in ILIAS) know(s) of alternatives here and want(s) to adapt the code in ilAuthProviderCAS even for ILIAS 11 to a newly proposed library via pull request, this would, of course, be enormously helpful for obtaining CAS authentication.

Given that the code was already broken in ILIAS 8 without any issue report in Mantis before mine, I doubt this is widely used. Nonetheless I will look for forums/mailing lists where I could spread the threat of removal in ILIAS 11.

**Comment by mjansenDatabay** [comment]
Thx for your contribution @Rillke !
